designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,312 @@
+      // +optional
+      Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
+}
+
+```
+#### Example
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    budgets:
+    - schedule: "* * * * *"
+      reason: [drifted, underutilized]
+      nodes: 10
+    # For all other reasons , only allow 5 nodes to be disrupted at a time
+    - nodes: 5
+      schedule: "* * * * *"
+
+```
+
+In the original proposed spec, karpenter allows the user to specify up to [50 budgets](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/apis/v1beta1/nodepool.go#L96)
+If there are multiple active budgets, karpenter takes the most restrictive budget. This same principle will be applied to the disruption budgets in this approach. The only difference in behavior is that each window will apply to list of reasons that are specifed rather than just all disruption methods. 
+### Pros + Cons 
+👍👍 Flexibility in Budget Allocation: Allows more flexibility in allocating budgets across multiple disruption reasons.
+👍👍 Reduced Configuration Complexity: Simplifies the configuration process, especially for similar settings across multiple reasons.
+👎 Potential for API Complexity: There might be confusion over whether the node count is shared between actions or if each action gets the node count individually. 
+
+Note some pros and cons between A + B can be shared, and are in a list next to the pros + cons for Single Reason Approach 
+
+#### Single Reason Approach: One Reason per Budget
+
+### Proposed Spec
+In this approach, each budget entry specifies a single reason for disruption.
+
+```go
+// Budget defines when Karpenter will restrict the
+// number of Node Claims that can be terminating simultaneously.
+type Budget struct {
+      // +kubebuilder:validation:Enum:={"consolidation","expired","empty","drift"}
+      // +optional
+      Reason string `json:"reason,omitempty" hash:"ignore"`
+      // Nodes dictates the maximum number of NodeClaims owned by this NodePool
+      // that can be terminating at once. This is calculated by counting nodes that
+      // have a deletion timestamp set, or are actively being deleted by Karpenter.
+      // This field is required when specifying a budget.
+      // This cannot be of type intstr.IntOrString since kubebuilder doesn't support pattern
+      // checking for int nodes for IntOrString nodes.
+      // Ref: https://github.com/kubernetes-sigs/controller-tools/blob/55efe4be40394a288216dab63156b0a64fb82929/pkg/crd/markers/validation.go#L379-L388
+      // +kubebuilder:validation:Pattern:="^((100|[0-9]{1,2})%|[0-9]+)$"
+      // +kubebuilder:default:="10%"
+      Nodes string `json:"nodes" hash:"ignore"`
+      // Schedule specifies when a budget begins being active, following
+      // the upstream cronjob syntax. If omitted, the budget is always active.
+      // Timezones are not supported.
```

Comment:
I can already taste the sadness of some poor American engineer having to update the disruption policies twice a year for DST.